xds/internal/xdsclient/xdsresource/cluster_resource_type.go

```diff
@@ -0,0 +1,171 @@
+		return &DecodeResult{Name: name, Resource: &ClusterResourceData{Resource: ClusterUpdate{}}}, err
+	}
+
+	return &DecodeResult{Name: name, Resource: &ClusterResourceData{Resource: cluster}}, nil
+
+}
+
+// ClusterResourceData wraps the configuration of a Cluster resource as received
+// from the management server.
+//
+// Implements the ResourceData interface.
+type ClusterResourceData struct {
+	ResourceData
+
+	// TODO: We have always stored update structs by value. See if this can be
+	// switched to a pointer?
+	Resource ClusterUpdate
+}
+
+// Equal returns true if other is equal to r.
+func (l *ClusterResourceData) Equal(other ResourceData) bool {
+	if l == nil && other == nil {
+		return true
+	}
+	if (l == nil) != (other == nil) {
+		return false
+	}
+	return proto.Equal(l.Resource.Raw, other.Raw())
+
+}
+
+// ToJSON returns a JSON string representation of the resource data.
+func (l *ClusterResourceData) ToJSON() string {
+	return pretty.ToJSON(l.Resource)
+}
+
+// Raw returns the underlying raw protobuf form of the cluster resource.
+func (l *ClusterResourceData) Raw() *anypb.Any {
+	return l.Resource.Raw
+}
+
+// ClusterResourceWatcher wraps the callbacks invoked by the xDS client
+// implementation for different events corresponding to the listener resource
+// being watched.
+type ClusterResourceWatcher interface {
+	// OnResourceChanged is invoked when an update for the resource being
+	// watched is received from the management server. The ResourceData
+	// parameter needs to be type asserted to the appropriate type for the
+	// resource being watched.
+	OnResourceChanged(*ClusterResourceData)
+
+	// OnError is invoked under different error conditions including but not
+	// limited to the following:
+	//	- authority mentioned in the resource is not found
+	//	- resource name parsing error
+	//	- resource deserialization error
+	//	- resource validation error
+	//	- ADS stream failure
+	//	- connection failure
+	OnError(error)
+
+	// OnResourceDoesNotExist is invoked for a specific error condition where
+	// the requested resource is not found on the xDS management server.
+	OnResourceDoesNotExist()
+}
+
+type delegatingClusterWatcher struct {
+	watcher ClusterResourceWatcher
+}
+
+func (d *delegatingClusterWatcher) OnGenericResourceChanged(data ResourceData) {
+	l := data.(*ClusterResourceData)
+	d.watcher.OnResourceChanged(l)
+}
+
+func (d *delegatingClusterWatcher) OnError(err error) {
+	d.watcher.OnError(err)
+}
+
+func (d *delegatingClusterWatcher) OnResourceDoesNotExist() {
+	d.watcher.OnResourceDoesNotExist()
+}
+
+// WatchCluster uses xDS to discover the configuration associated with the
+// provided cluster resource name.
+//
+// This is a convenience wrapper around the WatchResource() API and callers are
+// encouraged to use this over the latter.
+func WatchCluster(c XDSClient, resourceName string, watcher ClusterResourceWatcher) (cancel func()) {
```

Comment:
A global function that accepts an object that is called 1:1 -- would it not be better to have these as methods on the `XDSClient`, still?

Reply:
As mentioned in the other comment, if we can find a nice name for the first parameter here, it wouldn't look so bad.

The problem with making it a method on `XDSClient` is:
- It is an interface, and if we want to add support for new resource types, we would have to make breaking changes to the API
- Also, if we want to add support for a new resource type only in google3, it would painful to add that method to this interface

Reply:
> It is an interface, and if we want to add support for new resource types, we would have to make breaking changes to the API

Well, it's an internal interface, and one nobody else should implement even if it wasn't, so that's probably not a big concern.

> Also, if we want to add support for a new resource type only in google3, it would painful to add that method to this interface

For non-"standard" resource types, there would always be a generic watch method.


But I also missed that this was a different package, so maybe just ignore this right now.

Reply:
> Well, it's an internal interface, and one nobody else should implement even if it wasn't, so that's probably not a big concern.

Given that there is interest in the community for this interface to be moved out of `xds/internal`, it would be nice to keep the interface simple as long as we can and avoid potential breaking changes.

Reply:
I'm not sure we'll ever want to actually do that, as it will hamper our ability to make changes.